[Woo POS][Surveys] Hook survey notification scheduler with order creation and POS

iamgabrielma opened the pull request:
Closes WOOMOB-1498
Closes WOOMOB-1457

## Description
This PR adds the missing requirements to the `PointOfSaleNotificationScheduler` to schedule/fire POS survey notifications:
- Potential merchants will be sent 1 notification 1 minute after they create an order.
- Current merchants will be sent 1 notification 5 minutes after they open POS, once they've already used it at least once.
- Both are non-exclusive, a merchant can receive both notifications, just once each.

## Changes
- Differentiates between "current" and "potential" merchant notifications when we store them to app settings.
- Updates `PointOfSaleNotificationScheduler` to take into account if the notification has already been scheduled or not for each case
- Schedules `.potentialMerchant` if needed once a merchant creates an order
- Schedules `.currentMerchant` if needed once a merchant re-opens POS a second time

## Testing information
- Update `PointOfSaleNotificationScheduler.timeIntervalInSeconds` to something like 5 seconds.
- On a physical device, with notifications turned on, and in a US/UK store:
- Session 1: Create an order in IPP, no need for payment. You should see the notification.
- Session 1:Load POS, you should see no notification.
- Re-run the app
- Session 2: Create another order in IPP, you should see no notification
- Session 2: Load POS, you should see notification
- Re-run the app
- Session 3: Load POS, you should see no notification

There's a `resetPOSSurveyNotificationScheduled` action that can be set in the scheduler init if needed for testing, this will erase all the POS survey values stored in-app.

```
Task { @MainActor in
  let action = AppSettingsAction.resetPOSSurveyNotificationScheduled { _ in }
  stores.dispatch(action)
}
```

iamgabrielma committed:
get and set in-app values for potential merchant case when checking eligibility

iamgabrielma committed:
add new actions to separate potential from current merchant value

iamgabrielma committed:
generate new models copiable

iamgabrielma committed:
handle notification based on if has been sent yet, or has opened POS

iamgabrielma committed:
hook notification scheduler in order creation and POS entry point

iamgabrielma committed:
make test target compile

iamgabrielma committed:
update tests after injecting stores to scheduler

iamgabrielma committed:
lint

iamgabrielma commented on WooCommerce/Classes/POS/TabBar/POSTabCoordinator.swift:
```diff
@@ -116,11 +116,21 @@ final class POSTabCoordinator {
 private extension POSTabCoordinator {
+    func schedulePOSSurveyNotificationIfNeeded() {
+        Task { @MainActor in
+            await PointOfSaleNotificationScheduler().scheduleLocalNotificationIfEligible(for: .currentMerchant)
```
In the next PR we'll inject this one and add further tests.

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -1025,6 +1025,9 @@ final class EditableOrderViewModel: ObservableObject {
             guard let self else { return }
             self.onFinished(order)
             self.trackCreateOrderSuccess(usesGiftCard: usesGiftCard)
+            Task {
+                await PointOfSaleNotificationScheduler().scheduleLocalNotificationIfEligible(for: .potentialMerchant)
+            }
```
In the next PR we'll inject this one and add further tests

staskus replied:
For now, this never gets called, likely because `EditableOrderViewModel.swift` gets deinitialized.

iamgabrielma commented on WooCommerce/WooCommerceTests/POS/POSNotificationSchedulerTests.swift:
```diff
@@ -3,21 +3,26 @@ import UserNotifications
+    private let mockStores: MockStoresManager
 
-    init() {
+    init() async throws {
```
Interestingly, tests were crashing due to the action dispatcher until I made this init async ( as well as the struct tagged as MainActor).

staskus replied:
I think I noticed that Woo app stores require to be dispatched on the main thread.

iamgabrielma replied:
Yes, the dispatcher requires that any action runs on the main thread, otherwise we'd crash, so I marked the struct as `@MainActor` thinking it should suffice, but it seems that this is not enough, and I also had to mark the init as async throws. I'm not fully convinced of AI explanation of why:

> _Marking the struct as MainActor doesn’t automatically make the initializer synchronous-safe. The initializer runs before the instance is fully isolated to the main actor. This means if your initializer body touches any main-actor–isolated state or calls functions that are MainActor, Swift needs to hop to the main actor before completing initialization — and that’s an asynchronous operation._

We do seem to use the same init in other POS-related suites, so I'm guessing is some gotcha with SwiftTesting as well.

staskus commented on Modules/Sources/Yosemite/Stores/AppSettingsStore.swift:
```diff
@@ -293,10 +293,18 @@ public class AppSettingsStore: Store {
-            setPOSSurveyNotificationScheduled(onCompletion: onCompletion)
-        case .getPOSSurveyNotificationScheduled(onCompletion: let onCompletion):
-            getPOSSurveyNotificationScheduled(onCompletion: onCompletion)
+        case .setPOSSurveyPotentialMerchantNotificationScheduled(onCompletion: let onCompletion):
```
Although there's nothing inherently bad with it, it's a bit cumbersome and overengineered. A lot of line changes with unclear value. I wonder why we have this store over simple `UserDefaults`: e.g `boolFor(key: .posSurveyPotentialMerchantNotificationScheduled)`. So much boilerplate required to add a simple check 😀.  All the tests are also verifying the same thing over and over...

How was your experience with this code?

iamgabrielma replied:
> How was your experience with this code?
> So much boilerplate required to add a simple check

I agree with it being cumbersome. 

When the initial specs shared the flag state for both surveys wasn't that noticeable, but as soon as we introduce the second flag and the check for POS becomes a bit too much. While I followed the existing design of other flags and surveys, I agree that we should revisit the approach to use a simpler `UserDefaults` (or other) approach rather than dispatching from stores. I've opened a task on WOOMOB-1519 for the moment.

staskus commented on Modules/Tests/YosemiteTests/Stores/AppSettingsStoreTests.swift:
```diff
@@ -1558,43 +1558,183 @@ extension AppSettingsStoreTests {
+        XCTAssertFalse(result)
+    }
+
+    func test_getPOSSurveyCurrentMerchantNotificationScheduled_returns_true_after_setting_as_scheduled() throws {
```
Hopefully, AI typed this and saved you time 😀 

For some reason, it triggered me a bit that we make numerous async calls with completion blocks that we have to wait for to check the value of a boolean.

staskus commented on Modules/Sources/Yosemite/Actions/AppSettingsAction.swift:
```diff
@@ -369,15 +369,31 @@ public enum AppSettingsAction: Action {
+
+    /// Gets whether POS mode has been opened at least once
+    ///
+    case getHasPOSBeenOpenedAtLeastOnce(onCompletion: (Bool) -> Void)
```
super-nit; it could be shortened a bit, although it reminds me of old-school `Obj-C` conventions that almost read like a book https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html

```suggestion
    case getHasPOSBeenOpened(onCompletion: (Bool) -> Void)
```

staskus commented on WooCommerce/WooCommerceTests/POS/POSNotificationSchedulerTests.swift:
```diff
@@ -177,6 +194,72 @@ struct PointOfSaleNotificationSchedulerTests {
 
+    @Test func scheduleLocalNotificationIfEligible_when_potentialMerchant_already_scheduled_then_currentMerchant_can_still_be_scheduled() async throws {
+        // Given
+        let siteSettings = sampleSiteSettings(countryCode: "US")
```
👍 clear tests

staskus commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -1025,6 +1025,9 @@ final class EditableOrderViewModel: ObservableObject {
             guard let self else { return }
             self.onFinished(order)
             self.trackCreateOrderSuccess(usesGiftCard: usesGiftCard)
+            Task {
+                await PointOfSaleNotificationScheduler().scheduleLocalNotificationIfEligible(for: .potentialMerchant)
+            }
```
Task itself could hold it while it's being executed.
```suggestion
            let scheduler = PointOfSaleNotificationScheduler()
            Task { [scheduler] in
                await scheduler.scheduleLocalNotificationIfEligible(for: .potentialMerchant)
            }
```

staskus reviewed (approved):
Overall is good. I'm not blocking since it's behind the feature flag.

❌ Scheduling is not called for potential merchant, likely because EditableOrderViewModel gets released
✅ Scheduling works for existing merchant, all the checks to verify that it was already shown work as well.

Thanks!

iamgabrielma committed:
rename to use POS prefix

mimics 5eafca4b7f9243d5a261bcdad9936c38b6b5c489 , for some reason the merge conflicts from this rename in the base branch aren’t being resolved correctly, so redoing the same commit here.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/257">23.5</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Merge branch 'trunk' into task/WOOMOB-1498-pos-surveys-complete-eligibility-checks

# Conflicts:
#	WooCommerce/Classes/POS/POSNotificationScheduler.swift
#	WooCommerce/WooCommerce.xcodeproj/project.pbxproj
#	WooCommerce/WooCommerceTests/POS/POSNotificationSchedulerTests.swift

iamgabrielma commented:
Thanks for the review @staskus !

> ❌ Scheduling is not called for potential merchant, likely because EditableOrderViewModel gets released

I cannot replicate this one: If you tested on split view/tablet, did you tap on `Create` button in the top right corner, or just `Recalculate`? (as we need to tap twice in tablets to create it). When tapping `Create` we should still have a hold of the EditableOrderViewModel:

<img width="1494" height="396" alt="Screenshot 2025-10-16 at 13 11 46" src="https://github.com/user-attachments/assets/2c439cde-353c-4b0a-aee4-9f460e45247c" />

https://github.com/user-attachments/assets/283f0188-8c8d-45ff-8d08-dc2fd9070ff8

staskus commented:
> I cannot replicate this one: If you tested on split view/tablet, did you tap on Create button in the top right corner, or just Recalculate? (as we need to tap twice in tablets to create it). When tapping Create we should still have a hold of the EditableOrderViewModel:

I tested on iPad Air M2 26.0 device. @iamgabrielma , I tapped "Create". I can retest if needed, but I tried a few times, with and without holding a reference to Task, and both had different results.

iamgabrielma commented:
> > I cannot replicate this one: If you tested on split view/tablet, did you tap on Create button in the top right corner, or just Recalculate? (as we need to tap twice in tablets to create it). When tapping Create we should still have a hold of the EditableOrderViewModel:
> 
> I tested on iPad Air M2 26.0 device. @iamgabrielma , I tapped "Create". I can retest if needed, but I tried a few times, with and without holding a reference to Task, and both had different results.

Gotcha, I'll give it some more testing to see if I can make it fail before merging, as well as test your suggestion of using the capture list to hold the task 🤔

staskus commented:
@iamgabrielma I think we're good. I cannot repro properly anymore. Task should be holding the view model until Task has finished running.
